### PR TITLE
Inform about silent supergroup migrations

### PR DESCRIPTION
**Pull request by Madnex**
I think it would be worth mentioning that telegram groups are migrated automatically behind the scenes to supergroups under certain conditions. If not aware this can lead to some unexpected behavior of the session plugin.

### CODE REVIEW

**Review by KnorpelSenf — COMMENTED** [review]
Good idea (we should mention this problem) but flawed execution. We should
- offer a solution or a workaround
- explain why no solution exists

Yes, that's a contradiction, and yes, that's the reason why we didn't explain this earlier.

The two chats are completely separate, so what are we going to do about this?

(Also, I'd like to improve on structure, style and orthography before merging—but we need to answer the above question first, anyway.)

**Comment by Madnex** [comment]
I believe if we want to share the same session data for both chats it is required to somehow store the link between the two chat ids. However, to my knowledge this link is only temporarily provided by telegram. The [message type](https://core.telegram.org/bots/api#message) has two properties that contain the link info `migrate_to_chat_id` and `migrate_from_chat_id`. 

This in combination with the `chat.id` of the update gives the link between the old group and the new super group. When I experimented with the migration behavior I could only get these properties (`migrate_to_chat_id` and `migrate_from_chat_id`) once though. Don't know how this is designed but for sure you can not access this information for all future messages.

Once this information is catched by the bot I think one solution could be to add a value (e.g. `migrated_to`) for the old session object (old key) that contains the new key. And additionally a new session object (with the new key) with the information for the old session (e.g. `migrated_from`). And then always when loading session data check for the `migrated_from` field and point to that one instead, i.e. the new session object is only a pointer towards the old session.

What do you think? Or would there be a reason to maintain two sessions for both chat ids and then apply some other logic to combine them later on?

btw: should we continue this here or create an issue for this discussion?

**Comment by KnorpelSenf** [comment]
There is a way to link one chat to the other, which is what you outlined. However, this doesn't mean that automatic session migrations from one chat to the other are possible. This is because new messages can be received by the bot from both chats concurrently for a short period of time. This leads to data races which cannot be fixed without help from the bot developer. That's what I meant when I said that we have to

> * explain why no solution exists

**Comment by KnorpelSenf** [comment]
How should we proceed with this?

**Comment by Madnex** [comment]
Do you have more official information on this? I can try to put this conversation into a comprehensive text, explaining the two bullet points. But I can't find much info on this topic so I'm not certain about what exactly to explain.

**Comment by KnorpelSenf** [comment]
Here is the Bot API dev himself confirming the data races: https://t.me/tdlibchat/44915

**Comment by Madnex** [comment]
> Here is the Bot API dev himself confirming the data races: https://t.me/tdlibchat/44915

Thanks for the info. Do I understand correctly that we can hope that this will be fixed eventually (https://t.me/tdlibchat/42933)?

**Comment by KnorpelSenf** [comment]
> > Here is the Bot API dev himself confirming the data races: https://t.me/tdlibchat/44915
> 
> Thanks for the info. Do I understand correctly that we can hope that this will be fixed eventually (https://t.me/tdlibchat/42933)?

No. Read the respective message thread, it continued with https://t.me/tdlibchat/44906.

**Comment by Madnex** [comment]
Okay, I see. Added some more explanation about this problematic with two possible (inconsistent workarounds). Should we put more details or keep it on this level?

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
I think this level of detail is good, I like your changes.

There are still a few things that could be improved, though:

1. You describe the issue in a warning box, and then start a new section to suggest the workarounds. It makes more sense to explain the problem and the solution in one. Perhaps we could name the section just Supergroup Migrations or something, and then do all the explaining.
2. Do you think it would better in its own section, rather than putting it alongside the session keys? On the one side it is related to session keys, but on the other side it's still somewhat of a rare edge case. It makes sense to provide readers with the important information first (how to connect to a storage?), and when their bot is running, we can tell them about these edge cases to harden the code base (what if there are chat migrations?). I'm a little undecided here, what do you think?
3. You currently mention two workarounds: doing unreliable chat migrations, or disabling sessions for regular groups. There are more ways to go about this:
    - Ignore the problem. The bot's session data will effectively reset when a group is migrated. Simple, reliable, default behaviour, but potentially unexpected once per chat.
    - Only storing temporary data (or data with timeouts) in the session, and using a database for the important things that need to be migrated when a chat migrates. This can then use transactions and custom logic to handle concurrent data access from the old and the new chat. A lot of effort and has a performance cost, but the only truly reliable way to solve this problem.
    - Letting the users decide explicitly. (“This chat was migrated, do you want to transfer the bot data?”) Much more reliable and transparent than automatic migrations due to the artificially added delay, but worse UX.
4. Try keeping it at one sentence per line, as it is explained in the contribution guide in this repo. (There might be a few more nitpicks about style later on because for some reason we all care a bit too strongly about excellent language here, but it won't be much effort on your side, so don't worry about that for now.)


Thanks for your efforts! Looking forward to getting this ready for merge :)

**Comment by Madnex** [comment]
Thanks for the detailed feedback! 

For 1. and 2.: Yes, this is something I was wondering as well. I think it's worth making it a section. For the placement within the documentation: Is there any page / section about general caveats / good to know things? Something a bit like in the "Advanced/[Deployment Checklist](https://grammy.dev/advanced/deployment.html#deployment-checklist)" but more general. If not, maybe it's worth opening a new page? Or just in the [Overview page](https://grammy.dev/advanced/#overview-advanced-topics)? 

Do you have other caveats / tips in mind that are worth noting? Maybe also just general Telegram behavior that one might stumble over while developing a bot. This could all go to the decided place then.

And then in the [Session Page](https://grammy.dev/plugins/session.html#session-keys) we just put a short warning box linking to the explanation.

The suggested other workarounds are added in the new commit. If the content is fine like this, it is now just a matter of finding a place for it and prettifying the language / style :)

File: site/docs/plugins/session.md
Comment (KnorpelSenf) [reply site/docs/plugins/session.md:339]: @roj1512 what's your opinion on this?

Comment (Madnex) [reply site/docs/plugins/session.md]: I added two sentences in [#61b72f1](https://github.com/grammyjs/website/pull/559/commits/61b72f14fdbec4f4a20be58606056a9e123f8974). Maybe this puts it into context a bit?

Comment (KnorpelSenf) [reply site/docs/plugins/session.md]: Looking good, thanks

**Comment by roj1512** [comment]
Nothing is remaining except my take over.

**Comment by KnorpelSenf** [comment]
Awesome!

**Comment by Madnex** [comment]
> @Madnex could you fix the formatting issues?

What exactly do I need to change? I don't see where the format breaks?

**Comment by KnorpelSenf** [comment]
Check the CI task

**Comment by KnorpelSenf** [comment]
You can run the formatting script to fix it automatically :)

**Comment by KnorpelSenf** [comment]
See https://github.com/grammyjs/website#checking-correct-formatting-of-docs

**Comment by Madnex** [comment]
Thanks, got it I think 👍

**Comment by KnorpelSenf** [comment]
Awesome, thanks!

**Review by KnorpelSenf — APPROVED** [review]
LGTM, ready for translation IMO

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:331]:
```suggestion
If you are using sessions for groups, you should be aware that Telegram migrates regular groups to supergroups under certain circumstances (e.g. [here](https://github.com/telegramdesktop/tdesktop/issues/5593)).
```

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:333]:
```suggestion
This migration only occurs once for each group, but it can cause inconsistencies.
```

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:334]:
```suggestion
This is because the migrated chat is technically a completely different chat that has a different chat identifier, and hence a different session identifier.
```

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:336]:
```suggestion
Currently, there is no safe solution for this problem because messages from the two chats are also differently identified.
```

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:343]:
```suggestion
  For example, if a migration happens while a user is in a conversation powered by the [conversations plugin](./conversations.md), the conversation will be reset.
```

File: site/docs/plugins/session.md
Comment (roj1512) [inline site/docs/plugins/session.md:331]: How is this Telegram Desktop issue related?

Comment (Madnex) [reply site/docs/plugins/session.md:331]: It's just the only example / documentation for the impact of the migration I could find online. It's just an example of how to distinguish supergroups from normal groups (the part about "Actual behaviour"). Nothing in particular about Telegram Desktop.

Comment (roj1512) [reply site/docs/plugins/session.md:331]: Still a little misleading, don’t you think?

Comment (Madnex) [reply site/docs/plugins/session.md:331]: Well, I don't know. Tbh I did not even notice that this is about Telegram Desktop. We could also use [this link](https://stackoverflow.com/questions/54555821/why-telegram-group-automatically-becomes-super-group) instead

Comment (roj1512) [reply site/docs/plugins/session.md:331]: Can you remove the link?

Comment (KnorpelSenf) [reply site/docs/plugins/session.md:331]: It's documented at https://core.telegram.org/bots/api#message, too. Given how this is a main document that grammY is based on, it could make sense to change the link to that. Given how we constantly repeat information that is found on this page without naming it again, we might just as well remove the link. You decide :)

Comment (Madnex) [reply site/docs/plugins/session.md:331]: Okay, then I'll just remove the link and leave it to the reader :P 
Don't have access to my computer right now, will update it at the end of the week 👍

Comment (roj1512) [reply site/docs/plugins/session.md:331]: @Madnex I think you can just rest in your weekend. Thanks for the suggestion, @KnorpelSenf.

Comment (KnorpelSenf) [reply site/docs/plugins/session.md:331]: Awesome, thanks Roj. I can offer to review.

**Review by roj1512 — COMMENTED** [review]
LGTM. Thank you so much!

**Comment by habemuscode** [comment]
https://github.com/grammyjs/website/pull/628

**Review by CikiMomogi — APPROVED** [review]

File: site/docs/plugins/session.md
Comment (CikiMomogi) [inline site/docs/plugins/session.md:354]:
```suggestion
- Another workaround would be to limit the bot only for supergroups with [filtering](../guide/filter-queries.md) (or limit only session related features to supergroups).
```

File: site/docs/id/plugins/session.md
Comment (CikiMomogi) [inline site/docs/id/plugins/session.md:355]:
```suggestion
- Solusi lainnya adalah memanfaatkan [filter](../guide/filter-queries.md) untuk membatasi bot supaya bisa digunakan di supergroup saja, atau bisa juga membatasi fitur yang terkait dengan session hanya untuk supergroup.
```

File: site/docs/es/plugins/session.md
Comment (CikiMomogi) [inline site/docs/es/plugins/session.md:354]:
```suggestion
- Otra solución alternativa sería limitar el bot solo para los supergrupos con [filtrado](../guide/filter-queries.md) (o limitar solo las funciones relacionadas con la sesión a los supergrupos).
```

**Review by CikiMomogi — COMMENTED** [review]
A little touch...

**Pull request merged**
